### ISSUE TITLE
Alternate method of configuring holiday lists that adds more features

### ISSUE THREAD

**Issue by DeftNerd** [kylekatarnls/business-day#2]
It would take significant refactoring, but it might be nice to be able to add more information within the holiday configuration files to make it easier to display the actual purpose of a business being closed on a specific date and to override specific default dates.

The other benefit of the following proposal is that additional attributes could be added in the future without any breaking changes as long as the setHoliday parser just skips unrecognized attributes. 

Here is an example of what I'm proposing using the US holiday file:

```
<?php

return array(
'newyearsday' => [
  'date' => Carbon::parse("January 1 $year")->format('d/m'),
  'holiday' => [
    'en' => 'New Year\'s Day',
  ],
  'observed' => true,
],
'mlkday' => [
  'date' => Carbon::parse("third Monday of January $year")->format('d/m'),
  'holiday' => [
    'en' => 'Martin Luther King Jr. Day',
  ],
  'observed' => true,
],
'easter' => [
  'date' => Carbon::createFromDate($year, 3, 21)->addDays(easter_days($year))->format('d/m'),
  'holiday' => [
    'en' => 'Easter Sunday',
  ],
  'observed' => false,
],
'memorialday' => [
  'date' => Carbon::parse("last Monday of May $year")->format('d/m'),
  'holiday' => [
    'en' => 'Memorial Day',
  ],
  'observed' => true,
],
'independenceday' => [
  'date' => Carbon::parse("July 4 $year")->format('d/m'),
  'holiday' => [
    'en' => 'Independence Day',
  ],
  'observed' => true,
],
'laborday' => [
  'date' => Carbon::parse("first Monday of September $year")->format('d/m'),
  'holiday' => [
    'en' => 'Labor Day',
  ],
  'observed' => true,
],
'thanksgiving' => [
  'date' => Carbon::parse("fourth Thursday of November $year")->format('d/m'),
  'holiday' => [
    'en' => 'Thanksgiving',
  ],
  'observed' => true,
],
'christmasday' => [
  'date' => Carbon::parse("December 25 $year")->format('d/m'),
  'holiday' => [
    'de' => 'Weihnachtstag',
    'en' => 'Christmas Day',
    'es' => 'Día de Navidad',
    'fr' => 'Le jour de Noël',
  ],
  'observed' => true,
]);
```

```isHoliday()``` can continue to return a boolean value, but another function can be created such as ```isHolidayDetails()``` that would return a more expanded object or string of the holiday name. 

Perhaps a boolean can be passed such as ```isHoliday(true)``` that would cause the function to return the string of the holiday name if true, or false if false. 

For example:
```
Carbon::setHolidaysRegion('us-national');
Carbon::parse('2018-01-15')->isHoliday(true) // 'Martin Luther King Jr. Day' for 2018
Carbon::parse('2018-01-21')->isHoliday(true) // false for 2018
Carbon::parse('2019-01-15')->isHoliday(true) // false for 2019
Carbon::parse('2019-01-21')->isHoliday(true) // 'Martin Luther King Jr. Day' for 2019
Carbon::isHoliday(true) // (str) 'Holiday Name' if today is a holiday, (boolean) false if it is not a holiday
```

The purpose of using a 'slug' key for the array items is to enable a feature similar to ```ignoreHoliday('christmas')``` or ```observeHoliday('arborday')```. It would be a great way to use default national holidays while still allowing easy overrides for businesses that might be open on some of those days. 

As an aside, Easter was hell to calculate until I saw the php easter_days() function. :-)

**Comment by DeftNerd** [comment]
Also, is there a reason that the setHolidays function can't handle Carbon objects and instead expects a string formatted as 'd/m'?

**Comment by kylekatarnls** [comment]
Yes, there is a reason, it's more common. `m/d` is US-only, rest of the world use `d/m` (some examples: https://docs.oracle.com/cd/E19455-01/806-0169/overview-7/index.html)

A possible replacement could be `m-d` which match a part of ISO 8601, but as there is no standard for day and month date format only, I took the most used.

For the holiday's details. I'm not sure it should be the responsibility of this library. I hope to get contributions to complete holidays lists, but if it's needed to also translate holidays names in many languages, it could be quite discouraging for contributors.

But we can adding ids to holidays is a great idea, it would also allow to do:
```php
Carbon::setHolidays('us-my-business', array_filter(
    Carbon::getHolidays('us-national'),
    function ($date, $name) {
        return !in_array($name, array('christmas'));
    }
));
```
So holidays could be as easily be added or removed, and so ignore/overse would not be needed.

Then methods starting with `is` should always return a boolean, so if we add a method to retrieve data, it should be a getter. It can simply be `getHolidayId` which would return `'newyearsday'` for example.

Then this id could be used by the any app to get any name or additional data needed.

So if you're OK to add getHolidayId and ids in holidays files, PR's are welcome.

**Comment by DeftNerd** [comment]
It wasn't specifically the order of month/day or day/month, it was a question more of why the data being passed around between functions was being passed as a string rather than an actual Carbon object.

I'll try to whip something up some small PR's for you to consider.

**Comment by kylekatarnls** [comment]
It seems to me a complete Carbon instance would be overkill to just return month and day. Users could be tempted to set year or even hours, with string month-day format, there is no ambiguity about what is taken into account. Year is the input, day/month is the output, and it provides a clean short syntax to define holidays that are not year-dependent.

Thanks, I'll wait for your PR.

**Comment by kylekatarnls** [comment]
Hi @DeftNerd, do you still plan to propose a PR for this?

**Comment by kylekatarnls** [comment]
Hi again, a note about optimization. Such structure will have a cost:
```php
'holiday' => [
    'de' => 'Weihnachtstag',
    'en' => 'Christmas Day',
    'es' => 'Día de Navidad',
    'fr' => 'Le jour de Noël',
  ],
```
Supposing we get contributions to translate them in 40 languages. It will load a very big array in memory for most often only use the date.

Also, it's very rare to display multiple languages in the same interface. So the most efficient way would be to group what is typically loaded together.

OK to use string keys for holidays definition, it will allow to identify days across different regions and other systems.

Then such properties like `observed` are region-specific, when name/description are language-specific so it would be stored in 1 file per language and not linked to any region.

**Comment by kylekatarnls** [comment]
The `getHolidayId` will be added in the next release and is a beginning for this feature as it allows to identify the holidays:
https://github.com/kylekatarnls/business-day#getholidayid

**Comment by kylekatarnls** [comment]
Fully implemented in 1.1.0:
https://github.com/kylekatarnls/business-day/releases/tag/1.1.0

I will document all that in the README soon.
